### PR TITLE
Add "full grid" cell list with a separate list for each cell of a rectangular domain

### PR DESCRIPTION
**Pull request by efaulhaber**
Based on #12.

### CODE REVIEW

File: src/cell_lists/full_grid.jl [hunk 1]
```diff
@@ -0,0 +1,63 @@
+struct FullGridCellList{C, LI, MC}
```
Comment (svchb) [inline src/cell_lists/full_grid.jl:1]: Documentation is missing

File: src/nhs_grid.jl [hunk 2]
```diff
@@ -48,24 +48,25 @@ since not sorting makes our implementation a lot faster (although less paralleli
...
     cell_size           :: NTuple{NDIMS, ELTYPE} # Required to calculate cell index
-    cell_buffer         :: Array{NTuple{NDIMS, Int}, 2} # Multithreaded buffer for `update!`
+    cell_buffer         :: CB                    # Multithreaded buffer for `update!`
     cell_buffer_indices :: Vector{Int} # Store which entries of `cell_buffer` are initialized
     threaded_nhs_update :: Bool
 
     function GridNeighborhoodSearch{NDIMS}(search_radius, n_particles;
                                            periodic_box_min_corner = nothing,
                                            periodic_box_max_corner = nothing,
+                                           cell_list = DictionaryCellList{NDIMS}(),
```
Comment (svchb) [inline src/nhs_grid.jl:64]: Documentation is missing for `cell_list`

File: src/PointNeighbors.jl [hunk 3]
```diff
@@ -16,5 +16,6 @@ include("nhs_neighbor_lists.jl")
 export for_particle_neighbor, foreach_neighbor
 export TrivialNeighborhoodSearch, GridNeighborhoodSearch, PrecomputedNeighborhoodSearch
 export initialize!, update!, initialize_grid!, update_grid!
+export DictionaryCellList, FullGridCellList
```
Comment (svchb) [inline src/PointNeighbors.jl:19]: `DictionaryCellList` is also missing documentation

**Review by svchb — CHANGES_REQUESTED** [review]

File: src/cell_lists/full_grid.jl [hunk 4]
```diff
@@ -0,0 +1,117 @@
+"""
+    FullGridCellList(; min_corner, max_corner, search_radius = 0.0, periodicity = false)
+
+A simple cell list implementation where each (empty or non-empty) cell of a rectangular
+(axis-aligned) domain is assigned a list of points.
+This cell list only works when all points are inside the specified domain at all times.
+
+Use the default arguments to create an empty "template" cell list that can be used to create
```
Comment (LasNikas) [inline src/cell_lists/full_grid.jl:8]: there are no default arguments for `min_corner` and `max_corner`?

File: src/nhs_grid.jl [hunk 5]
```diff
@@ -184,33 +190,38 @@ end
...
-@inline function mark_changed_cell!(neighborhood_search, cell, coords_fun)
+@inline function mark_changed_cell!(neighborhood_search, cell_index, coords_fun)
     (; cell_list, cell_buffer, cell_buffer_indices) = neighborhood_search
 
-    for point in cell_list[cell]
-        if cell_coords(coords_fun(point), neighborhood_search) != cell
+    for point in cell_list[cell_index]
+        cell = cell_coords(coords_fun(point), neighborhood_search)
+
+        # `cell` is a tuple, `cell_index` is the linear index used internally be the
```
Comment (LasNikas) [inline src/nhs_grid.jl:215]:
```suggestion
        # `cell` is a tuple, `cell_index` is the linear index used internally by the
```

File: src/nhs_grid.jl [hunk 6]
```diff
@@ -184,33 +190,38 @@ end
...
+        # cell list to store cells inside `cell`.
+        # These can be identical (see `DictionaryCellList`)
```
Comment (LasNikas) [inline src/nhs_grid.jl:217]:
```suggestion
        # These can be identical (see `DictionaryCellList`).
```

**Review by LasNikas — CHANGES_REQUESTED** [review]

File: src/cell_lists/full_grid.jl (hunk 4 above)
Comment (efaulhaber) [reply src/cell_lists/full_grid.jl:8]: No, these are always mandatory. I've updated the docs to make that clear.

**Review by svchb — APPROVED** [review]

**Review by LasNikas — APPROVED** [review]

**Pull request merged**